src/betterproto/__init__.py

```diff
@@ -519,23 +519,67 @@ def __post_init__(self) -> None:
             if meta.group:
                 group_current.setdefault(meta.group)
 
-            if getattr(self, field_name) != PLACEHOLDER:
-                # Skip anything not set to the sentinel value
+            if self.__raw_get(field_name) != PLACEHOLDER:
+                # Found a non-sentinel value
                 all_sentinel = False
 
                 if meta.group:
                     # This was set, so make it the selected value of the one-of.
                     group_current[meta.group] = field_name
 
-                continue
-
-            setattr(self, field_name, self._get_field_default(field_name))
-
         # Now that all the defaults are set, reset it!
         self.__dict__["_serialized_on_wire"] = not all_sentinel
         self.__dict__["_unknown_fields"] = b""
         self.__dict__["_group_current"] = group_current
 
+    def __raw_get(self, name: str) -> Any:
+        return super().__getattribute__(name)
+
+    def __eq__(self, other):
+        if type(self) is not type(other):
+            return False
+
+        equal = True
+        for field_name in self._betterproto.meta_by_field_name:
+            self_val = self.__raw_get(field_name)
+            other_val = other.__raw_get(field_name)
+            if self_val is PLACEHOLDER and other_val is PLACEHOLDER:
+                continue
+            elif self_val is PLACEHOLDER:
+                self_val = self._get_field_default(field_name)
+            elif other_val is PLACEHOLDER:
+                other_val = other._get_field_default(field_name)
+
+            if self_val != other_val:
+                equal = False
+                break
+
+        return equal
+
+    def __repr__(self):
+        parts = [self.__class__.__name__, "("]
+        found = False
+        for field_name in self._betterproto.meta_by_field_name:
+            value = self.__raw_get(field_name)
+            if value is PLACEHOLDER:
+                continue
+            found = True
+            parts.extend([field_name, "=", repr(value), ", "])
+
+        if found:
+            parts.pop()
+        parts.append(")")
+        return "".join(parts)
```

Comment:
```suggestion
    def __repr__(self) -> str:
        parts = []
        for field_name in self._betterproto.meta_by_field_name:
            value = self.__raw_get(field_name)
            if value is PLACEHOLDER:
                continue
            parts.append(f"{field_name}={value!r}")

        return f"{self.__class__.__name__}({', '.join(parts)})"
```
Is a lot more readable

Reply:
Yes this is an improvement.